src/adapter/src/catalog/migrate.rs

```diff
@@ -182,6 +209,576 @@ pub(crate) async fn migrate(
+///
+/// Third we migrate existing single-output `CREATE SOURCE` statements.
+/// This includes existing Kafka and single-output load-generator
+/// subsources. This will generate an additional `CREATE TABLE .. FROM SOURCE`
+/// statement that copies over all the export-specific options. This table will use
+/// to the existing source statement's persist shard but use a new GlobalID.
+/// The original source statement will be updated to remove the export-specific options,
+/// renamed to `<original_name>_source`, and use a new empty shard while keeping its
+/// same GlobalId.
+///
+fn ast_rewrite_sources_to_tables(
+    tx: &mut Transaction<'_>,
+    now: NowFn,
+) -> Result<(), anyhow::Error> {
+    use maplit::btreemap;
+    use maplit::btreeset;
+    use mz_persist_types::ShardId;
+    use mz_proto::RustType;
+    use mz_sql::ast::{
+        CreateSourceConnection, CreateSourceOptionName, CreateSourceStatement,
+        CreateSubsourceOptionName, CreateSubsourceStatement, CreateTableFromSourceStatement, Ident,
+        KafkaSourceConfigOptionName, LoadGenerator, MySqlConfigOptionName, PgConfigOptionName,
+        RawItemName, TableFromSourceColumns, TableFromSourceOption, TableFromSourceOptionName,
+        UnresolvedItemName, Value, WithOptionValue,
+    };
+    use mz_storage_client::controller::StorageTxn;
+    use mz_storage_types::sources::load_generator::LoadGeneratorOutput;
+    use mz_storage_types::sources::SourceExportStatementDetails;
+    use prost::Message;
+
+    let items_with_statements = tx
+        .get_items()
+        .map(|item| {
+            let stmt = mz_sql::parse::parse(&item.create_sql)?.into_element().ast;
+            Ok((item, stmt))
+        })
+        .collect::<Result<Vec<_>, anyhow::Error>>()?;
+    let items_with_statements_copied = items_with_statements.clone();
+
+    // Any GlobalId that should be changed to a new GlobalId in any statements that
+    // reference it. This is necessary for ensuring downstream statements (e.g.
+    // mat views, indexes) that reference a single-output source (e.g. kafka)
+    // will now reference the corresponding new table, with the same data, instead.
+    let mut changed_ids = BTreeMap::new();
+
+    for (mut item, stmt) in items_with_statements {
+        match stmt {
+            // Migrate each `CREATE SUBSOURCE` statement to an equivalent
+            // `CREATE TABLE .. FROM SOURCE` statement.
+            Statement::CreateSubsource(CreateSubsourceStatement {
+                name,
+                columns,
+                constraints,
+                of_source,
+                if_not_exists,
+                mut with_options,
+            }) => {
+                let raw_source_name = match of_source {
+                    // If `of_source` is None then this is a `progress` subsource which we
+                    // are not migrating as they are not currently relevant to the new table model.
+                    None => return Ok(()),
```

Comment:
Shouldn't this be a `continue`?  Otherwise you'll exit out of the entire migration on the first progress subsource.

Reply:
Yes, wow 🤦🏽  good catch!

Reply:
this would have surfaced in my next PR where we enforce during planning that no old syntax is used, which is why I didnt catch it here! I guess we could also add tests that use `SHOW .. ` to verify the statements actually changed, rather than just verifying that queries still work on the same named objects

Reply:
Yeah, it would be nice to have a test that would have caught this, but if it would have definitely been caught by the next PR, then that's fine too.

Reply:
Yep - @nrainer-materialize might be adding checks to the test workflow he added to verify the actual `SHOW CREATE SOURCE` and `SHOW CREATE TABLE FROM SOURCE ..` statements have been updated, which would then catch this